### PR TITLE
feat: conversion of html generated by html-webpack-plugin

### PR DESCRIPTION
**Pull request by konpeki622**
Related to #83 

### transform functional webpack config with AST

  - The `webpackConfigure` and `chainWebpack` options could be defined as object or function in `vue.config.js` module
  - To avoid calling error when they were functional, I initialize config options and generate a temporary file (`vue.temp.config.js`) to reconfig `html-webpack-plugin`

### AST

- `FindWebpackConfigProperties`: initialize config options according to functional `webpackConfigure`
- `FindHtmlPluginChain`: parse `chainWebpack`
- `ChainWebpackTransformation`:  generate a temporary file (`vue.temp.config.js`) to reconfig `html-webpack-plugin`
- `FindHtmlConfigProperties`: reparse `chainWebpack` and initialize options for html plugin

### `html-webpack-plugin` is supported

  - Options will be applied to plugin `vite-plugin-html`
  - Variables injected to `index.html` will be transformed. for example, `<%= htmlWebpackPlugin.options.title %>` -> `<%= title %>`
  - Import `injectHtml` and `minifyHtml` from `html-webpack-plugin`

support these options:

- [x] title
- [x] template
- [x] favicon
- [x] minify
- [x] publicPath
- [x] templateContent
- [x] templateParameter

***
Thank you for your review. I will add test use cases.

### CODE REVIEW

File: src/ast-parse/astParse.ts [hunk 1]
```diff
@@ -71,105 +86,174 @@ export async function astParseRoot (
...
+        : path.resolve(replacedRootDir, 'vue.temp.config.js')
+      const vueConfig = await parseVueCliConfig(vueConfigPath)
+
+      // vueConfig.configureWebpack
+      let webpackConfig: Configuration = {}
+      if (vueConfig.configureWebpack && typeof vueConfig.configureWebpack !== 'function') {
+        webpackConfig = vueConfig.configureWebpack
+      } else if (vueConfig.configureWebpack) {
+        try {
+          webpackConfig = applyAstParsingResultToConfig(webpackConfig, 'FindWebpackConfigProperties', parsingResults)
```
Comment (Chieffo2021) [inline src/ast-parse/astParse.ts:154]: Whether `FindWebpackConfigProperties` can be imported from `constants` file.

File: src/transform/transformVuecli.ts [hunk 2]
```diff
@@ -26,12 +32,60 @@ export class VueCliTransformer implements Transformer {
...
+      const vueConfigPath = existsSync(path.resolve(rootDir, 'vue.temp.config.ts')) ? path.resolve(rootDir, 'vue.temp.config.ts') : path.resolve(rootDir, 'vue.temp.config.js')
+      const vueConfig = await parseVueCliConfig(vueConfigPath)
+
+      let webpackConfig: Configuration = {}
+      // vueConfig.configureWebpack
+      if (vueConfig.configureWebpack && typeof vueConfig.configureWebpack !== 'function') {
+        webpackConfig = vueConfig.configureWebpack
+      } else if (vueConfig.configureWebpack && astParsingResult) {
+        try {
+          webpackConfig = applyAstParsingResultToConfig(webpackConfig, 'FindWebpackConfigProperties', astParsingResult.parsingResult)
```
Comment (Chieffo2021) [inline src/transform/transformVuecli.ts:44]: Whether `FindWebpackConfigProperties` can be imported from `constants` file.

**Review by Chieffo2021 — CHANGES_REQUESTED** [review]

File: src/ast-parse/transformations/chainWebpackTransformation.ts [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  }
+  if (transformationParams.config.projectType === 'webpack') {
+    return null
+  }
+  if (!/vue.config.(js|ts)$/.test(fileInfo.path)) {
+    return null
+  }
+
+  const rootDir: string = transformationParams.config.rootDir
+  const extension: string = (/\.([^.]*)$/.exec(fileInfo.path) || [])[0]
```
[the remark this answers is not in the document]
Comment (konpeki622) [reply src/ast-parse/transformations/chainWebpackTransformation.ts:30]: That is uncorrect when the extension was longer than 1.

File: src/utils/astUtils.ts [hunk 4]
```diff
@@ -31,3 +31,89 @@ export function parseScriptSfc (fileInfo: FileInfo, lang?: string) : any {
   const scriptAST = jscodeshiftParser(fileInfo.source)
   return scriptAST
 }
+
+export function parseIdentifierFromBodyNodes (bodyNodes: any[], paramName: string): ParsingResultProperty[][] {
```
[the remark this answers is not in the document]
Comment (konpeki622) [reply src/utils/astUtils.ts]: I added comment to this function. It's a little hard to test.

File: src/ast-parse/astParse.ts [hunk 5]
```diff
@@ -67,74 +83,141 @@ export async function astParseRoot (
   const parsingResults: ParsingResult = {}
   const transformationResults: AstTransformationResult = {}
 
-  const transformationParams: TransformationParams = {
-    config: config
-  }
-  cliInstance.setTotal(cliInstance.total + resolvedPaths.length)
```
Comment (Chieffo2021) [inline src/ast-parse/astParse.ts:73]: We should not delete this line, the total should be plus the number of ast parsing times the number of files.

**Comment by Chieffo2021** [comment]
Now we have 22 automatic conversions
https://github.com/originjs/webpack-to-vite/blob/fa9a79ace9bcf938bb5c7e32d0d05b6fcaf2663f/src/cli/cli.ts#L50

**Review by Chieffo2021 — APPROVED** [review]

**Pull request merged**